### PR TITLE
task done

### PR DESCRIPTION
**Pull request by Rommelua**

### CODE REVIEW

File: src/main/java/com/internet/shop/controller/RegistrationController.java [hunk 1]
```diff
@@ -36,7 +37,9 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
         String passwordRep = req.getParameter("pswRep");
 
         if (password.equals(passwordRep)) {
-            User user = new User(name, login, password, Set.of(Role.of("USER")));
+            byte[] salt = HashUtil.getSalt();
+            String hashPassword = HashUtil.hashPassword(password, salt);
+            User user = new User(name, login, hashPassword, salt, Set.of(Role.of("USER")));
```
Comment (DmGrt) [inline src/main/java/com/internet/shop/controller/RegistrationController.java]: Maybe better to move this logic into create method in UserServiceImpl class

File: src/main/java/com/internet/shop/util/HashUtil.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        SecureRandom random = new SecureRandom();
+        byte[] salt = new byte[16];
+        random.nextBytes(salt);
+        return salt;
+    }
+
+    public static String hashPassword(String password, byte[] salt) {
+        StringBuilder hashedPassword = new StringBuilder();
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance("SHA-512");
```
Comment (DmGrt) [inline src/main/java/com/internet/shop/util/HashUtil.java]: Make "SHA-512" as constant

File: src/main/java/com/internet/shop/controller/InjectDataController.java [hunk 3]
```diff
@@ -24,13 +25,29 @@
...
-        userService.create(new User("admin", "admin", "1111", Set.of(Role.of("ADMIN"))));
-        userService.create(new User("User A", "userA", "1111", Set.of(Role.of("USER"))));
-        userService.create(new User("User B", "userB", "1111", Set.of(Role.of("USER"))));
-        userService.create(new User("User C", "userC", "1111", Set.of(Role.of("USER"))));
-        shoppingCartService.create(new ShoppingCart(3L));
-        shoppingCartService.create(new ShoppingCart(4L));
-        shoppingCartService.create(new ShoppingCart(5L));
+        User admin = new User("admin", "admin", "1111",
+                HashUtil.getSalt(), Set.of(Role.of("ADMIN")));
+        String password = HashUtil.hashPassword(admin.getPassword(), admin.getSalt());
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controller/InjectDataController.java]: I agree with Dmytro - let's move this logic in UserServiceImpl class

File: src/main/java/com/internet/shop/controller/RegistrationController.java (hunk 1 above)
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controller/RegistrationController.java]: I agree, let's not change this class

File: src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java [hunk 4]
```diff
@@ -14,9 +14,11 @@
 
     @Override
     public User login(String login, String password) throws AuthenticationException {
-        Optional<User> optionalUser = userService.getByLogin(login);
-        if (optionalUser.isPresent() && optionalUser.get().getPassword().equals(password)) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java:18]: we could call a private method in the right part of the statement in order not to throw the exception twice

File: src/main/java/com/internet/shop/util/HashUtil.java (hunk 2 above)
Comment (KseniiaMakarova) [reply src/main/java/com/internet/shop/util/HashUtil.java]: I agree, let's make it a constant

File: src/main/java/com/internet/shop/util/HashUtil.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            for (byte b: digest) {
+                hashedPassword.append(String.format("%02x", b));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/util/HashUtil.java]: let's add some informative message

File: src/main/resources/init_db.sql [hunk 6]
```diff
@@ -25,6 +25,7 @@ CREATE TABLE public.users
     login character varying(256) NOT NULL,
     password character varying(256) NOT NULL,
     deleted boolean NOT NULL DEFAULT false,
+    salt bytea,
```
Comment (KseniiaMakarova) [inline src/main/resources/init_db.sql]: is this the correct syntax?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/internet/shop/util/HashUtil.java (hunk 2 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/util/HashUtil.java]: done

File: src/main/java/com/internet/shop/util/HashUtil.java (hunk 5 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/util/HashUtil.java]: done

File: src/main/resources/init_db.sql (hunk 6 above)
Comment (Rommelua) [reply src/main/resources/init_db.sql]: It was generated by PGAdmin...

File: src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java (hunk 4 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java:18]: changed

File: src/main/java/com/internet/shop/controller/RegistrationController.java (hunk 1 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/controller/RegistrationController.java]: done

Comment (Rommelua) [reply src/main/java/com/internet/shop/controller/RegistrationController.java]: done

File: src/main/java/com/internet/shop/controller/InjectDataController.java (hunk 3 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/controller/InjectDataController.java]: done

File: src/main/resources/init_db.sql (hunk 6 above)
Comment (KseniiaMakarova) [inline src/main/resources/init_db.sql]:
```suggestion
    salt bytea NOT NULL,
```

File: src/main/java/com/internet/shop/util/HashUtil.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        StringBuilder hashedPassword = new StringBuilder();
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance(ALGORITHM);
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            for (byte b: digest) {
+                hashedPassword.append(String.format("%02x", b));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Can't hash password",e);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/util/HashUtil.java]:
```suggestion
            throw new RuntimeException("Can't hash password", e);
```

File: src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java [hunk 8]
```diff
@@ -14,9 +15,10 @@
...
-            return optionalUser.get();
+        Optional<User> user = userService.getByLogin(login);
+        if (user.isPresent() && HashUtil.hashPassword(password, user.get().getSalt())
+                .equals(user.get().getPassword())) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java]: let's create the private method `isPasswordValid(User user, String password)` to make this more readable

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/resources/init_db.sql (hunk 6 above)
Comment (Rommelua) [reply src/main/resources/init_db.sql]: done

File: src/main/java/com/internet/shop/util/HashUtil.java (hunk 7 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/util/HashUtil.java]: done

File: src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java (hunk 8 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/security/AuthenticationServiceImpl.java]: done

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**